### PR TITLE
Bug 1817075: MCC & MCO don't free leader leases during shut down -> 10 minutes of leader election timeouts

### PR DESCRIPTION
**Pull request by jkyros**
**TL;DR**: SIGTERM/SIGINT --> cancel `runContext` --> shutdown -->   cancel `leaderContext` --> release leader lease --> terminate 

**- What I did**
- added two contexts:
   - a `runContext` (for cencelling our running state)  and 
   - a `leaderContext` (for cancelling our leader lease) 
 to both the controller and the operator run functions
- added a signal handler that catches SIGINT/SIGTERM to controller and operator
- updated run functions such that: 
   - signal handler cancels `runContext` when SIGTERM/SIGINT is received
   - operator/controller shuts down when `runContext` is cancelled, and then sequentially cancels `leaderContext`
- updated leaderElection config in controller and operator to release leader lease when `leaderContext` is cancelled 
- shutdown metrics listener properly 

**- How to verify it**
- delete a controller or an operator pod 
- observe logging that shows it shutting down/releasing the lease 
- observe that the pod that replaces it has short ( < 1 minute) leader election wait 

**- Description for the changelog**
Controller and operator shutdown cleanly on normal termination and release their leader lease cleanly, resulting in faster leader election times, since we don't have to wait for the previous lease to time out anymore. 

Fixes: [BZ1817075](https://bugzilla.redhat.com/show_bug.cgi?id=1817075)

### CODE REVIEW

File: cmd/machine-config-operator/start.go [hunk 1]
```diff
@@ -91,20 +98,24 @@ func runStartCmd(cmd *cobra.Command, args []string) {
 
 		go controller.Run(2, ctrlctx.Stop)
 
-		select {}
+		// wait here in this function until the context gets cancelled (which tells us whe were being shut down)
+		<-ctx.Done()
+		glog.Infof("Context has returned done, waiting 10 seconds to let any in-progress work finish")
+		<-time.After(10 * time.Second)
```
Comment (cgwalters) [inline cmd/machine-config-operator/start.go]: Perhaps worth factoring this into a helper function to share with controller?

Or at least just a cookie for future readers to know there's code duplication, like
`// Note, if you're changing this you probably also want to change machine-config-controller/start.go`
or so?

File: cmd/machine-config-controller/start.go [hunk 2]
```diff
@@ -90,16 +94,16 @@ func runStartCmd(cmd *cobra.Command, args []string) {
...
-		<-ctx.Done()
-		glog.Infof("Context has returned done, waiting 10 seconds to let any in-progress work finish")
-		// TODO(jkyros): someday it would be nice to have this be event-driven so the controller could track and acknowledge that it had completed
-		// any in-progress work (e.g. it is in the middle of a sync), but for now just give it a little bit of time.
-		<-time.After(10 * time.Second)
+		<-runContext.Done()
+		glog.Infof("Context has returned done, waiting 5 seconds to let any in-progress work finish")
+		// TODO(jkyros): someday it might be nice to have the controller wait for any in-progress work (e.g. it is in the middle of a sync at this very moment),
+		// but for now just give it a little bit of time.
+		<-time.After(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (jkyros) [reply cmd/machine-config-controller/start.go:101]: Yeah,  this was my concern too. I talked about it with Jerry a little last week. 

 My thought here was: 
- if there are no more events coming in, and no workers working (because the context cancelled them all),  `syncMachineConfigPool` can't run again because there's no event to trigger it
- so the only thing we *should* have to worry about is the one `syncMachineConfigPool` that is already in progress

but yes it would make me feel safer if we had something to stop us if some race-contion-y thing happened and something snuck through the queue while everything was reacting to the context cancellation.

[the remark this answers is not in the document]
Comment (jkyros) [reply cmd/machine-config-controller/start.go:101]: Jerry and I talked some more about this. As for that variable, checking the context should be just as good, (if cancelled, quietly no-op), but say I add a context check at the beginning of `syncMachineConfigPool`:
-  that doesn't save us from a "half-done" sync (like if we were already in the middle of the function body)
-  if we wait for that "half-done" sync, our syncs _could_ be even slower than they are today?

[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply cmd/machine-config-controller/start.go:101]: Some other options we discussed:
- have the next controller, after acquiring the lease, add a small wait to ensure the previous pod is (most likely) terminated
- somehow track all the subcontrollers and routines and terminate them synchronously
- check what other controllers are doing for this

[the remark this answers is not in the document]
Comment (cgwalters) [reply cmd/machine-config-controller/start.go:101]: I agree we should check what other operators are doing here.

That said, to avoid races I think the general pattern is:

- Use a "done" channel we can poll in each controller sync loop
- [Use a sync.WaitGroup](https://pkg.go.dev/sync#WaitGroup) to wait for the completion of the polling goroutines

[the remark this answers is not in the document]
Comment (jkyros) [reply cmd/machine-config-controller/start.go:101]: I did read through the [CVO PR](https://github.com/openshift/cluster-version-operator/pull/424) you sent me, the CVO is really meticulous about it (see: https://github.com/openshift/cluster-version-operator/blob/master/pkg/start/start.go), taking an approach where they: 

- explicitly try to handle crashes in and reap all their goroutines and 
- have a graceful shutdown period (2 minutes) where they allow some time for that to happen 

I don't know how much "return on investment" we get from doing that if we don't really want to wait for things to shutdown, but if we think that's the approach we want to take, I can set ours up that way also?

[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply cmd/machine-config-controller/start.go:101]: Hmm, so in theory, the longest "sync" we have takes 10 minutes, which is the waiting for required pools loop. Now I think about it, if we do wait for "graceful shutdown" during upgrades, it will never happen for the first MCC pod since the required pools will never be ready in that 10 minute blocking poll.

Put another way, maybe a better way to approach this is to rework that polling loop to not be blocking, such that we could gracefully shutdown

[the remark this answers is not in the document]
Comment (cgwalters) [reply cmd/machine-config-controller/start.go:101]: For correctness, everything that's making a blocking call should take a context and be cancelled when that happens.  So given your example of `syncRequiredMachineConfigPools()` for example...we should be passing a `context` to that, then it gets passed down into the API calls like 
```
diff --git a/pkg/operator/sync.go b/pkg/operator/sync.go
index 69e4aa742..dd131f751 100644
--- a/pkg/operator/sync.go
+++ b/pkg/operator/sync.go
@@ -684,7 +684,7 @@ func (optr *Operator) syncRequiredMachineConfigPools(_ *renderConfig) error {
 				return false, nil
 			}
 			optr.setOperatorStatusExtension(&co.Status, lastErr)
-			_, err = optr.configClient.ConfigV1().ClusterOperators().UpdateStatus(context.TODO(), co, metav1.UpdateOptions{})
+			_, err = optr.configClient.ConfigV1().ClusterOperators().UpdateStatus(ctx), co, metav1.UpdateOptions{})
 			if err != nil {
 				lastErr = errors.Wrapf(lastErr, "failed to update clusteroperator: %v", err)
 				return false, nil
```

[the remark this answers is not in the document]
Comment (cgwalters) [reply cmd/machine-config-controller/start.go:101]: And there is also [PollWithContext](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollWithContext) instead of plain `wait.Poll` that we should use.

[the remark this answers is not in the document]
Comment (cgwalters) [reply cmd/machine-config-controller/start.go:101]: How about this as a short term plan:

As soon we lose the lease, call `os.Exit(0)` - do not pass Go¹, do not collect $200.

This means we could interrupt some of the controllers mid change, but they already need to be robust to that in theory.



¹ Get it?  We're not "passing" any more Go code, the process is exiting...

[the remark this answers is not in the document]
Comment (cgwalters) [reply cmd/machine-config-controller/start.go:101]: I can't say for sure myself...possibly/probably? Immediately calling `os.Exit(0)` would reinforce that that's what we expect, and be robust to someone coming along later and adding something after the leader election loss (as unlikely as that is probably).

(I don't have a really strong opinion on this bit)

But I do think we should not call `time.Sleep()` - we should exit as fast as possible.

**Comment by jkyros** [comment]
Alright, so along the lines of the conversation in https://github.com/openshift/machine-config-operator/pull/3185#discussion_r895956976, this simplifies this back to one single context and and gets rid of the delays so everything immediately exits. 
- This should be no worse than what we have today because we already don't gracefully shutdown, we just die to SIGTERM
- I put up https://github.com/openshift/machine-config-operator/pull/3194 to see how much work it would be to do this "correctly" and shut everything down properly, but only look if you're curious, I have no expectation that that goes in anytime soon

File: cmd/machine-config-controller/start.go [hunk 3]
```diff
@@ -82,20 +89,23 @@ func runStartCmd(cmd *cobra.Command, args []string) {
...
 	leaderElectionCfg := common.GetLeaderElectionConfig(cb.GetBuilderConfig())
 
-	leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
-		Lock:          common.CreateResourceLock(cb, startOpts.resourceLockNamespace, componentName),
-		LeaseDuration: leaderElectionCfg.LeaseDuration.Duration,
-		RenewDeadline: leaderElectionCfg.RenewDeadline.Duration,
-		RetryPeriod:   leaderElectionCfg.RetryPeriod.Duration,
+	leaderelection.RunOrDie(runContext, leaderelection.LeaderElectionConfig{
+		Lock:            common.CreateResourceLock(cb, startOpts.resourceLockNamespace, componentName),
+		ReleaseOnCancel: true,
```
Comment (yuqi-zhang) [inline cmd/machine-config-controller/start.go:100]: Am I correct in understanding that essentially, this release is the crux of the PR here. Previously, we took the leader election (lock?) and never released it until the timeout happens?

**Review by yuqi-zhang — APPROVED** [review]
Generally looks fine. With this new change, do you foresee any potential dangerous scenarios we can fall into?

File: cmd/machine-config-controller/start.go (hunk 3 above)
Comment (jkyros) [reply cmd/machine-config-controller/start.go:100]: You are correct.

**Comment by jkyros** [comment]
> Generally looks fine. With this new change, do you foresee any potential dangerous scenarios we can fall into?

I don't see any "concrete" danger given how things actually operate. I do see a "hypothetical" danger in that a verrrrry small amount of time could potentially elapse between when we cancel the context and when we terminate with os.Exit(). In that verrrrry small amount of time:
- whatever is in progress is still progressing (e.g. an in-progress `syncMachineConfigPool`) , *but* since we've also cancelled the controller context (which stops the informers, etc)  we aren't going to process any *new* events at that point. 

To expand on that: 

- A new pod can't take over and do work until we release the lease
- the same context that cancels the lease also cancels the controllercontext, so it can't receive any new informer events/take on any new work once we cancel 
- we die immediately after we release the lease 
- the only gap where anything can technically happen is between 
	- when the lease is cancelled and 
	- when we die to os.Exit(0)
- And that's in client-go and it fires the moment we leave the leaderelection Run() loop, which we do immediately when the context is cancelled. https://github.com/openshift/machine-config-operator/blob/37b741601f9b7ff9b2e1870102cc2970b24e1835/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go#L206-L213

So it's:
-  "can another pod grab the lease and start working on something we're already working on in the time it takes for a defer to trigger call our function that calls `os.Exit(0)` ?". 

And while technically I can say yes (because a very small amount of time could elapse during those go statements due to scheduling) realistically I think I can say no. Also, I'm pretty sure it will always take longer than that (the time between the context cancellation and `os.Exit()` ) for the next pod to acquire the leader lease. 

( the dual context/5 second wait I had in there before was  "try to get that in-progress work done before we exit" vs this now which is "try to exit before we get any more work done" )

**Comment by yuqi-zhang** [comment]
Ok I think we should be good to merge this. To be safe, I'm going to go ahead and

/payload 4.11 ci blocking

just to run a set of additional tests

**Review by yuqi-zhang — APPROVED** [review]
/lgtm

All green on the payload side, I think this is of high value to upgrade timings to get in for 4.11, so let's try to merge this.

Thanks for the great work John!
